### PR TITLE
Fix to issue 32972: Grow GArray multiplicatively

### LINKED ISSUE #32972: GArray should scale mulplicatively instead of additively
The current code just adds 64 to the array size each time we run out of space. We should consider scaling by 1.5x or something instead to ensure fewer allocations on large arrays.

### PR DESCRIPTION
**Pull request by mihir-b-shah**
Fixes #32972

This tries to fix this referenced issue.
It changes to the multiplicative growth.

I am a newcomer to open source so please tell me if I have not submitted something right, thank you!

### CODE REVIEW

**Review by lambdageek — CHANGES_REQUESTED** [review]
The `Makefile.am.in`->`Makefile.am` file change is not necessary.

In dotnet/runtime MonoVM doesn't use autotools anymore, we use `cmake`.

In mono/mono (where this PR is mirrored), the `Makefile.am.in` file is intentional - it's used by the packaging process when producing official Mono packages.

---

As for the substantive change in this PR, it looks good. We should keep an eye on the perf numbers after we merge this, @CoffeeFlux .

File: src/mono/mono/eglib/garray.c [hunk 1]
```diff
@@ -50,7 +50,7 @@ ensure_capacity (GArrayPriv *priv, guint capacity)
 	if (capacity <= priv->capacity)
 		return;
 	
-	new_capacity = (capacity + 63) & ~63;
+	new_capacity = ((capacity*3 >> 1) + 63) & ~63;
```
Comment (lambdageek) [inline src/mono/mono/eglib/garray.c:53]: might be better to do this computation as a shift and an add instead.

**Comment by CoffeeFlux** [comment]
Yep re: perf, cc: @naricc

**Comment by danmosemsft** [comment]
@mihir-b-shah could you please retitle the PR to suggest what the change is? that makes it easier when browsing PRs.

**Comment by mihir-b-shah** [comment]
Hi I changed the issue name and changed to a shift and add, I also reversed the makefile change.

**Comment by lambdageek** [comment]
@mihir-b-shah `Makefile.am` and the change to `.gitignore` looks like it's still in the PR.

You can probably do something like this:
git rm src/mono/mono/mini/Makefile.am
git reset c7fda3b94a50297e6ece9db4ddc79ed039394dd3 -- src/mono/mono/mini/.gitignore
git restore --source=c7fda3b94a50297e6ece9db4ddc79ed039394dd3 -- src/mono/mono/mini/.gitignore
git add -A src/mono/mono/mini
git commit -m "undo changes to mono/mini/Makefile.am"
```

**Comment by mihir-b-shah** [comment]
Hi I fixed it I think the Makefile.am is gone and it is restored to the gitignore.

**Review by lambdageek — APPROVED** [review]

**Comment by danmosemsft** [comment]
@mihir-b-shah FYI a tip, if you edit your top comment to change "#32972" to "Fixes #32972" then github will close the issue when this is merged.

**Comment by CoffeeFlux** [comment]
Just waiting on CI for the mono/mono mirror PR—I've restarted the failing lanes, so hopefully that should be done soon. Thanks for your contribution!

**Comment by naricc** [comment]
@CoffeeFlux We don't have the capacity to do a full perf run for every change, only the ones we expect to have a wide impact across multiple benchmarks. So I don't plan to do one of this change. I will monitor the weekly runs after this change though.

**Comment by CoffeeFlux** [comment]
Yeah I don't think we need a separate run for this, just wanted to make sure you're aware it exists since it likely has perf implications. 🙂

**Comment by lambdageek** [comment]
Thanks for the contribution @mihir-b-shah !

**Pull request merged**